### PR TITLE
Add postgres

### PR DESCRIPTION
**Pull request by Leaxe**

### CODE REVIEW

File: src/lib/stores/store.ts [hunk 1]
```diff
@@ -1,4 +1,6 @@
-import { writable } from "svelte/store";
+import { asyncWritable, writable, type Stores } from "@square/svelte-store";
+import _ from 'lodash';
+import {activityToActivityComponentsRow, type ActivitiesRow, type ActivityComponentsRow} from "$lib/activities/activities";
 
 export type AbstractActivity = {
     id: number;
```
Comment (silso) [inline src/lib/stores/store.ts]: Should we change these to an ID type?

File: src/lib/stores/store.ts [hunk 2]
```diff
@@ -22,42 +24,59 @@ export type FundamentalActivity = AbstractActivity & {
...
+        .map(activity => [_.omit(activity, 'components'), activityToActivityComponentsRow(activity)])
+        .unzip()
+        .value() as [ActivitiesRow[], ActivityComponentsRow[]] // shouldn't need to do this
+    return [activitiesRows, _.flatten(activityComponentsRows)]
+}
+
+// For some reason the `oldValues` argument in the second asyncWritable callback was always exactly the same as new,
+// So we're doing it manually
+const oldActivitiesStore = writable<Activity[]>([]);
+// This is used to debounce writing to the database
```
Comment (silso) [inline src/lib/stores/store.ts:9]: Explain how this works in the comment here, maybe

File: src/lib/stores/store.ts [hunk 3]
```diff
@@ -22,42 +24,59 @@ export type FundamentalActivity = AbstractActivity & {
...
-            components: undefined,
-        },
-    ]
+function activitiesToActivitiesAndActivityComponentsRows(activities: Activity[]): [ActivitiesRow[], ActivityComponentsRow[]] {
+    const [activitiesRows, activityComponentsRows] = _.chain(activities)
+        .map(activity => [_.omit(activity, 'components'), activityToActivityComponentsRow(activity)])
+        .unzip()
+        .value() as [ActivitiesRow[], ActivityComponentsRow[]] // shouldn't need to do this
+    return [activitiesRows, _.flatten(activityComponentsRows)]
+}
```
Comment (silso) [inline src/lib/stores/store.ts]: Should this and the Activity type definitions be moved to `$lib/activities/activities`? Actually they should maybe be moved to their own file in the store folder here

File: src/lib/stores/store.ts [hunk 4]
```diff
@@ -22,42 +24,59 @@ export type FundamentalActivity = AbstractActivity & {
...
+    },
+    async (newActivities) => {
+        if (await writing.load()) return
+        
+        writing.set(true);
+        setTimeout(async () => {
+            const oldActivities = await oldActivitiesStore.load();
+            
+            const [newActivitiesRows, newActivityComponentsRows] = activitiesToActivitiesAndActivityComponentsRows(newActivities)
+            const [oldActivitiesRows, oldActivityComponentsRows] = activitiesToActivitiesAndActivityComponentsRows(oldActivities ?? [])
```
Comment (silso) [inline src/lib/stores/store.ts]: remove the nullish coalescing

File: src/routes/api/activities/+server.ts [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+		} else {
+			const components = activityComponents.get(activity.id);
+			if (components === undefined) throw new TypeError;
+			return {
+				...activity,
+				fundamental: false,
+				components
+			}
+		}
+	})
```
Comment (silso) [inline src/routes/api/activities/+server.ts]: Next time comment better

**Comment by Leaxe** [comment]
im sorry but this cant wait another sprint, review it immediately

File: src/routes/api/activities/+server.ts [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+	return new Response(JSON.stringify(activities)); // FIXME: prolly dont want to stringify
+}) satisfies RequestHandler;
+
+export const POST = (async ({ request, locals }) => {
+	const { sql } = locals;
+	const {
+		updatedActivitiesRows,
+		updatedActivityComponentsRows,
+		removedActivitiesRows,
```
Comment (silso) [inline src/routes/api/activities/+server.ts:47]: I suppose this is what tRPS will help with?

File: src/routes/api/activities/+server.ts [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+			ON CONFLICT (parent, child) DO UPDATE
+				SET weight = EXCLUDED.weight
+		`],
+		// NOTE: these might not work at all
+		...(removedActivitiesRows.length === 0) ? [] : [sql`
+			DELETE FROM "Activities"
+			WHERE id IN ${sql(removedActivitiesRows)}
+		`],
+		...(removedActivityComponentsRows.length === 0) ? [] : [sql`
+			DELETE FROM "Activities"
```
Comment (silso) [inline src/routes/api/activities/+server.ts:75]: Make "Activities" a variable?

File: src/routes/api/activities/+server.ts [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+	const result = await sql.begin(sql => [
+		...(updatedActivitiesRows.length === 0) ? [] : [sql`
+			INSERT INTO "Activities" ${
+				sql(updatedActivitiesRows)
+			}
+			ON CONFLICT (id) DO UPDATE
+				SET name = EXCLUDED.name,
+					value = EXCLUDED.value,
+					description = EXCLUDED.description,
+					fundamental = EXCLUDED.fundamental
```
Comment (silso) [inline src/routes/api/activities/+server.ts:60]: Can this stuff be typed?

File: src/routes/+page.svelte [hunk 9]
```diff
@@ -1,6 +1,10 @@
...
+    import ActivityForm from '$lib/activities/ActivityForm.svelte';
 </script>
 
-<ActivityForm bind:activity={$activities[0]}/>
\ No newline at end of file
+{#await activities.load()}
+<p>Currently loading...</p>
+{:then}
+<ActivityForm bind:activity={$activities[0]} />
+{/await}
```
Comment (silso) [inline src/routes/+page.svelte:10]: Beautiful

File: src/app.d.ts [hunk 10]
```diff
@@ -1,9 +1,16 @@
 // See https://kit.svelte.dev/docs/types#app
 // for information about these interfaces
+
+import type {Sql} from "postgres";
```
Comment (silso) [inline src/app.d.ts:5]: yucky, move this down below the comment

**Review by silso — CHANGES_REQUESTED** [review]

**Review by silso — CHANGES_REQUESTED** [review]
change

File: src/routes/api/activities/+server.ts (hunk 8 above)
Comment (Leaxe) [reply src/routes/api/activities/+server.ts:60]: lol no not really

File: src/routes/api/activities/+server.ts (hunk 6 above)
Comment (Leaxe) [reply src/routes/api/activities/+server.ts:47]: maybe

File: src/lib/stores/store.ts (hunk 1 above)
Comment (Leaxe) [reply src/lib/stores/store.ts]: yes

Comment (Leaxe) [reply src/lib/stores/store.ts]: no

Comment (Leaxe) [reply src/lib/stores/store.ts]: string

File: src/lib/activities/ActivityForm.svelte [hunk 11]
```diff
@@ -1,5 +1,5 @@
 <script lang=ts>
-    import type { Activity } from "$lib/stores/store";
+	import type {Activity} from "./activities";
```
Comment (silso) [inline src/lib/activities/ActivityForm.svelte:2]: What's this??

**Review by silso — APPROVED** [review]

**Pull request merged**